File: testdata/precision-suite-perl/tp_known_fns.pl
#!/usr/bin/perl
# Honest false negatives: real flows a *correct* scanner should flag that nox's
# pragmatic Perl line-recognizer does NOT. They are annotated so they score as
# recall (a number below 1.0), never as silence. Samples are never edited to fake
# the score — the engine is built to catch what it honestly can, and the rest is
# recorded here. See README.md "Known gaps". Neither line below produces ANY
# finding, so each scores purely as a missed true positive (recall), never a
# false positive.
use strict;
use warnings;

# GAP 1 — taint laundered through a hash element. The tainted value is stored in
# $args{cmd}, but the assignment target is a subscripted lvalue, not a bare
# scalar. nox's engine tracks simple-identifier assignments only (no container /
# element sensitivity — the documented Python/JS/Ruby limit), so the taint is
# lost at the hash store and the sink read of $args{cmd} looks clean.
sub launder_hash {
    my %args;
    $args{cmd} = $ENV{CMD};
    system("run $args{cmd}"); # nox-expect: TAINT-002
}

# GAP 2 — cross-subroutine flow through a package global. The source lands in
# our $PAYLOAD in one sub and is read by a sink in another. nox's same-file
# interprocedural pass tracks LOCAL HELPER CALLS via function summaries, not
# shared package/global state, so a value laundered across two subs through a
# global is not joined. This is the documented boundary of the intraprocedural +
# local-summary model, not a Perl-specific defect.
our $PAYLOAD;

sub stash {
    $PAYLOAD = $ENV{DATA};
}

sub flush {
    system("logger $PAYLOAD"); # nox-expect: TAINT-002
}

1;

File: testdata/precision-suite-ruby/tp_known_fns.rb
# Honest false negatives: real vulnerabilities a CORRECT scanner should flag but
# nox's intraprocedural Ruby LINE recognizer does NOT catch yet. Each is
# annotated with the rule a correct scanner would fire, so it scores as a
# recall gap (FN) rather than being quietly omitted — the whole point of an
# honest measurement corpus. See README "Known gaps" for why each is missed.
class KnownGapsController
  # FN #1 — `render inline:` template injection. A tainted value in an inline
  # ERB template is real XSS/SSTI, but the recognizer keys sinks by call NAME and
  # cannot distinguish `render inline:` (dangerous) from `render plain:`/`json:`
  # (auto-escaped, safe). Firing on bare `render` over-fired the clean auto-
  # escaped renders, so the `render` sink was intentionally dropped. Result: this
  # genuine flow is missed.
  def template
    name = params[:name]
    render inline: "<h1>Hello #{name}</h1>" # nox-expect: TAINT-003
  end

  # FN #2 — cross-method flow through an instance variable. The source lands in
  # @cmd in one action and the sink reads @cmd in another. nox's same-file
  # interprocedural pass tracks LOCAL HELPER CALLS via summaries, not shared
  # object/instance state, so a taint laundered through an @ivar across two
  # methods is not joined.
  def capture
    @cmd = params[:cmd]
  end

  def execute_captured
    system @cmd # nox-expect: TAINT-002
  end

  # FN #3 — metaprogramming via Object#send. A tainted method name dispatched
  # through `send` is code injection, but the argument is a method-name string,
  # not a value flowing into a recognized sink shape; the line recognizer does
  # not model dynamic dispatch.
  def dispatch
    action = params[:action]
    target.send(action, params[:arg]) # nox-expect: TAINT-005
  end
end
